zkevm-circuits/src/sig_circuit/ecdsa.rs

```diff
@@ -81,39 +81,43 @@ where
     // load required constants
     let zero = scalar_chip.load_constant(ctx, FpConfig::<F, SF>::fe_to_constant(SF::zero()));
     let one = scalar_chip.load_constant(ctx, FpConfig::<F, SF>::fe_to_constant(SF::one()));
-    let neg_one = scalar_chip.load_constant(ctx, FpConfig::<F, SF>::fe_to_constant(-SF::one()));
+    let point_at_infinity = EcPoint::construct(
+        ecc_chip
+            .field_chip()
+            .load_constant(ctx, fe_to_biguint(&CF::zero())),
+        ecc_chip
+            .field_chip()
+            .load_constant(ctx, fe_to_biguint(&CF::zero())),
+    );
 
     // compute u1 = m * s^{-1} mod n
     let s2 = scalar_chip.select(ctx, &one, s, &s_is_zero);
     let u1 = scalar_chip.divide(ctx, msghash, &s2);
     let u1 = scalar_chip.select(ctx, &zero, &u1, &s_is_zero);
-    let u1_is_one = {
-        let diff = scalar_chip.sub_no_carry(ctx, &u1, &one);
-        let diff = scalar_chip.carry_mod(ctx, &diff);
-        scalar_chip.is_zero(ctx, &diff)
-    };
 
     // compute u2 = r * s^{-1} mod n
     let u2 = scalar_chip.divide(ctx, r, &s2);
     let u2 = scalar_chip.select(ctx, &zero, &u2, &s_is_zero);
 
-    // u3 = 1 if u1 == 1
-    // u3 = -1 if u1 != 1
-    // this ensures u1 + u3 != 0
-    let u3 = scalar_chip.select(ctx, &one, &neg_one, &u1_is_one);
+    // we want to compute u1*G + u2*PK, there are two edge cases
+    // 1. either u1 or u2 is 0; we use binary selections to handle the this case
+    // 2. or u1*G + u2*PK is an infinity point; this is computed with paddings
 
-    let u1_plus_u3 = scalar_chip.add_no_carry(ctx, &u1, &u3);
-    let u1_plus_u3 = scalar_chip.carry_mod(ctx, &u1_plus_u3);
-
-    // compute (u1+u3) * G
-    let u1u3_mul = fixed_base::scalar_multiply::<F, _, _>(
+    // =================================
+    // case 1:
+    // =================================
+    let u1_is_zero = scalar_chip.is_zero(ctx, &u1);
+    let u1_prime = scalar_chip.select(ctx, &one, &u1, &u1_is_zero);
```

Comment:
Is it possible to avoid `u1_is_zero` and simple do:
```suggestion
    let u1_prime = scalar_chip.select(ctx, &one, &u1, &s_is_zero);
```

Reply:
Possibly not? As it could be zero if `s != 0` but `m * s^{-1} mod n == 0 (mod n)`, Is that right?